lib/src/buffer.dart

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+import 'dart:async';
+
+/// Creates a [StreamTransformer] which collects values and emits when it sees a
+/// value on [trigger].
+///
+/// If there are no pending values when [trigger] emits, the next value on the
+/// source Stream will immediately flow through. Otherwise, the pending values
+/// are released when [trigger] emits.
+///
+/// Errors from the source stream or the trigger are immediately forwarded to
+/// the output.
+StreamTransformer<T, List<T>> buffer<T>(Stream trigger) => new _Buffer(trigger);
+
+List<T> _collectToList<T>(T element, List<T> soFar) {
+  soFar ??= <T>[];
+  soFar.add(element);
+  return soFar;
+}
+
+/// A StreamTransformer which aggregates values and emits when it sees a value
+/// on [_trigger].
+///
+/// If there are no pending values when [_trigger] emits the first value on the
+/// source Stream will immediately flow through. Otherwise, the pending values
+/// and released when [_trigger] emits.
+///
+/// Errors from the source stream or the trigger are immediately forwarded to
+/// the output.
+class _Buffer<T> implements StreamTransformer<T, List<T>> {
+  final Stream _trigger;
+
+  _Buffer(this._trigger);
+
+  @override
+  Stream<List<T>> bind(Stream<T> values) {
+    StreamController<List<T>> controller;
+    if (values.isBroadcast) {
+      controller = new StreamController<List<T>>.broadcast();
+    } else {
+      controller = new StreamController<List<T>>();
+    }
+
+    List<T> currentResults;
+    bool waitingForTrigger = true;
+    StreamSubscription valuesSub;
+    StreamSubscription triggerSub;
+
+    cancelValues() {
+      var sub = valuesSub;
+      valuesSub = null;
+      return sub?.cancel() ?? new Future.value();
+    }
+
+    cancelTrigger() {
+      var sub = triggerSub;
+      triggerSub = null;
+      return sub?.cancel() ?? new Future.value();
+    }
+
+    closeController() {
+      var ctl = controller;
+      controller = null;
+      return ctl?.close() ?? new Future.value();
+    }
+
+    emit() {
+      controller.add(currentResults);
+      currentResults = null;
```

Comment:
For safety I would just set `waitingForTrigger = true` here, even though it might already be in that state. I think you always want it to be true after this?